incompatibility/incompatibility-report/oas3-traverse.go

```diff
@@ -0,0 +1,160 @@
+package incompatibility
+
+import (
+	"reflect"
+)
+
+type TraverseError struct {
+	desrciptorString string
+}
+
+func (e TraverseError) Error() string {
+	return e.desrciptorString
+}
+
+// Structure for generic single-input single-output function with error indication
+type GenericOperation struct {
+	op         func(interface{}) (interface{}, error)
+	inputType  reflect.Type
+	outputType reflect.Type
+}
+
+type transverseComp int
+
+const (
+	DiffType = iota
+	SameType
+)
+
+func (op1 GenericOperation) typeEquality(op2 GenericOperation) transverseComp {
+	if op1.inputType == op2.inputType && op1.outputType == op2.outputType {
+		return SameType
+	}
+	return DiffType
+}
+
+// arg function should be a single-input single-output function
+// creates generic wrapper around function, indicates error on return
+func makeGenericOperation(function interface{}) (GenericOperation, error) {
+	switch inputType := reflect.TypeOf(function); inputType.Kind() {
+	case reflect.Func:
+		// Continue function
+		validFunction := inputType.NumIn() == 1 && inputType.NumOut() == 1
+		if validFunction {
+			funcInputT := inputType.In(0)
+			funcOutputT := inputType.Out(0)
+			formattedGenericFunction := func(input interface{}) (interface{}, error) {
+				switch t := reflect.TypeOf(input); t {
+				case funcInputT:
+					functionAsValue := reflect.ValueOf(function)
+					returnAsInterface := functionAsValue.Call([]reflect.Value{reflect.ValueOf(input)})[0].Interface()
+					returnType := reflect.TypeOf(returnAsInterface)
+					if returnType != funcOutputT {
+						return nil, TraverseError{"Bad Return Type in Invocation"}
+					}
+					return returnAsInterface, nil
+				default:
+					return nil, TraverseError{"Bad Input Type in Inovocation"}
+				}
+			}
+			return GenericOperation{
+				op:         formattedGenericFunction,
+				inputType:  funcInputT,
+				outputType: funcOutputT,
+			}, nil
+		}
+		goto failGen
+	default:
+		// Invalid input
+		goto failGen
+	}
+failGen:
+	return GenericOperation{}, TraverseError{"input is not a single-input single-output function"}
+}
+
+// function should be a single-input single-output function whose output type is *IncompatibilityReport
+// performs error checking for these requirements and casts into generic type
+func makeGenericIncompatibilityReportFunc(function interface{}) (func(interface{}) (*IncompatibilityReport, error), error) {
+	genGeneric, err := makeGenericOperation(function)
+	badType := genGeneric.outputType != reflect.TypeOf(&IncompatibilityReport{})
```

Comment:
I think you can type the argument to this function, instead of do this check at runtime?
`makeGenericIncompatibilityReportFunc(function func(interface {}) *IncompatibilityReport)`
I'm not 100% sure this will work...?